internal/provider/resource_gitlab_branch_test.go

```diff
@@ -0,0 +1,205 @@
+					testAccCheckGitlabBranchCommit("foo2"),
+				),
+			},
+			// Test ImportState
+			{
+				ResourceName:      "gitlab_branch.foo",
+				ImportState:       true,
+				ImportStateVerify: true,
+			},
+			// update properties in resource
+			{
+				Config: testAccGitlabBranchConfig(rInt2, project.PathWithNamespace),
+				Check: resource.ComposeTestCheckFunc(
+					testAccCheckGitlabBranchExists("foo", &branch, rInt2),
+					testAccCheckGitlabBranchExists("foo2", &branch2, rInt2),
+					testAccCheckGitlabBranchAttributes("foo", &branch, &testAccGitlabBranchExpectedAttributes{
+						Name:    fmt.Sprintf("testbranch-%d", rInt2),
+						CanPush: true,
+						Commit:  true,
+					}),
+					testAccCheckGitlabBranchAttributes("foo2", &branch2, &testAccGitlabBranchExpectedAttributes{
+						Name:    fmt.Sprintf("testbranch2-%d", rInt2),
+						CanPush: true,
+						Commit:  true,
+					}),
+					testAccCheckGitlabBranchRef("foo", "main"),
+					testAccCheckGitlabBranchRef("foo2", fmt.Sprintf("testbranch-%d", rInt2)),
+				),
+			},
+		},
+	})
+}
+
+func testAccCheckGitlabBranchCommit(n string) resource.TestCheckFunc {
+	return func(s *terraform.State) error {
+		rs := s.RootModule().Resources[fmt.Sprintf("gitlab_branch.%s", n)]
+		commit := rs.Primary.Attributes["commit.0.id"]
+		if commit == "" {
+			return fmt.Errorf("expected commit to be populated")
+		}
+		return nil
+	}
+}
+
+func testAccCheckGitlabBranchRef(n, expectedRef string) resource.TestCheckFunc {
+	return func(s *terraform.State) error {
+		rs := s.RootModule().Resources[fmt.Sprintf("gitlab_branch.%s", n)]
+		ref := rs.Primary.Attributes["ref"]
+		if ref != expectedRef {
+			return fmt.Errorf("expected ref: %s got: %s", expectedRef, ref)
+		}
+		return nil
+	}
+}
+
+func testAccCheckGitlabBranchDestroy(s *terraform.State) error {
+	for _, rs := range s.RootModule().Resources {
+		if rs.Type != "gitlab_branch" {
+			continue
+		}
+		name := rs.Primary.Attributes["name"]
+		project := rs.Primary.Attributes["project"]
+		branch, resp, err := testGitlabClient.Branches.GetBranch(project, name)
+		if err == nil {
+			if branch != nil && branch.Name == name {
+				return fmt.Errorf("Branch still exists")
+			}
+		}
+		if resp.StatusCode != 404 {
+			return err
+		}
+		return nil
+	}
+	return nil
+}
+
+func testAccCheckGitlabBranchAttributes(n string, branch *gitlab.Branch, want *testAccGitlabBranchExpectedAttributes) resource.TestCheckFunc {
+	return func(s *terraform.State) error {
+		rs, ok := s.RootModule().Resources[fmt.Sprintf("gitlab_branch.%s", n)]
+		if !ok {
+			return fmt.Errorf("Not Found: %s", n)
+		}
+
+		_, branchName, err := parseTwoPartID(rs.Primary.ID)
+		if err != nil {
+			return errors.New("Error in splitting project and branch IDs")
+		}
+		if branch.WebURL == "" {
+			return errors.New("got empty web url")
+		}
+		if branchName != want.Name || branchName != branch.Name {
+			return fmt.Errorf("got name %s; want %s", branch.Name, want.Name)
+		}
```

Comment:
I'm probably forgetting some context, but why parse the ID? Shouldn't we always expect the branch name to equal the `want.Name`? How about this:

```suggestion
		if branch.WebURL == "" {
			return errors.New("got empty web url")
		}
		if branch.Name != want.Name {
			return fmt.Errorf("got name %s; want %s", branch.Name, want.Name)
		}
```